### DISCUSSION TITLE
Support 8.0+ #[Test] attributes

### DISCUSSION THREAD

**Discussion by mimre25** [olimorris/neotest-phpunit#27] [Ideas]
With PHP 8.0 [attributes](https://docs.phpunit.de/en/10.5/attributes.html) were introduced and can be used to annotate tests.

I have a larger code base that mixes the traditional (`/** @test */`) and the modern (`#[Test]`) style of annotating test functions.

I've dug a bit into the `neotest-phpunit` codebase[^1] but I'm hitting a wall either with the treesitter query, or the lack of its result being passed around.

This is the query that I came up with:
```lisp
    ( 
     (method_declaration
      (attribute_list
        (attribute_group
            (attribute) @test_attribute (#match? @test_attribute "Test")
        )
      )
      (
        (name) @test.name
      )
     ) @test.definition
    )
```

This neatly matches the test, however, the problem stems from the attribute (`#[Test]`) being considered part of the function declaration, ultimately leading to an off-by-one when it comes to line-number for the test ids.

```php
<?php

namespace Tests\Feature\Api;

use PHPUnit\Framework\Attributes\Test;
use Tests\TestCase;

class DemoTest extends TestCase
{
    #[Test] // is considered part of the method declaration
    public function impersonationCanReturnADemoToken() //PHPUnit output uses this linenumber though
    {}
}
?.
```
I'd love to contribute, but I'm just hitting a wall and would need a pointer to get past it.

[^1]: I'm also implementing Laravel Sail support right now (which would probably also cover docker). Hopefully I can create a PR for this soon :slightly_smiling_face:

**Comment by olimorris** [comment]
What about trying:

```scm
((method_declaration
  (attribute_list) @attribute (#eq? @attribute "#[Test]")
))
```

Just had a quick play with the following code:

```php
<?php

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    public function test_that_true_is_true()
    {
        $this->assertTrue(true);
    }
    
    #[Test]
    public function it_does_something(): void
    {
        $this->assertTrue(true);
    }
}
```

and it feels like it's on the right line. Writing Tree-sitter queries is a real trial and error!

If I get some time in the week I'll try and look into it.

**Reply by mimre25** [reply]
That query still includes the line containing `#[Test]` for me.

**Reply by olimorris** [reply]
Yeah that wasn't the answer just a query that selected the test based on attribute. You can chain additional queries based on that and should be able to get there.

Are you using the online playground?

**Reply by mimre25** [reply]
I'm using the playground inside neovim. 

I tried to come up with a query, but the best I could get is the one above.
I really don't know how I can select "starting the next line" with a treesitter query, and if I just try to select the siblings of the attribute, the capture will always just select a single one.

```
    ( 
     (method_declaration
      (attribute_list
        (attribute_group
            (attribute) @test_attribute (#match? @test_attribute "Test")
        )
      )
      (
        (name) @test.name
        (_)+
      )  @test.definition ;; this now only captures a single sibling of name instead of the whole group
     )
    )
```

As I said: I really hit a wall here, and I've played around quite a bit :grimacing:

**Reply by olimorris** [reply]
I'm about 90% of the way there with this query:

```scm
((method_declaration
      (attribute_list (_) @attribute (#eq? @attribute "#[Test]"))
      name: (name) @test.name
    )) @test.definition
```

It's just selecting the `#[Test]` attribute as part of the test_definition. I can't workout if we actually want that and therefore need to adapt the logic of the plugin.

**Reply by olimorris** [reply]
So with this code:

```php
<?php

use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;

class AttributeTest extends TestCase
{
    #[Test]
    public function it_should_pass_if_attributes_work()
    {
        $this->assertTrue(true);
    }
}
```

The issue is that my query, it's picking up two @test.name values. Both `#[Test]` and `it_should_pass_if_attributes_work`!

**Reply by mimre25** [reply]
I've done some googling and it seems like that a treesitter capture can only contain a single node.
That means, we can only capture the whole `method_declaration` including the attribute.

Now I can see two solutions for this:
1. Adapt the code to read the files, checking for the `#[Test]` attribute and fixing the off-by-one
2. Just capturing some other node that represents the first line of the test (eg the `visibility_modifier`)

I've tested (2) as it seems pretty simple, and it works. 
However, I don't know if there are any nasty side-effects of just saying the keyword `public` is the `@test.definition` in the query.


```
    ( 
     (method_declaration
      (attribute_list
        (attribute_group
            (attribute) @test_attribute (#match? @test_attribute "Test")
        )
      )
      (
        (visibility_modifier) @test.definition    
        (name) @test.name
      )
     ) 
    )
```

**Reply by olimorris** [reply]
Fantastic! That works brilliantly. We may need to add `TestDox` support at some point but thanks so much for your efforts. Great to keep this plugin moving with the times.

**Reply by mimre25** [reply]
`TestDox` alone doesn't define a test, so it would need any of the other "identifiers" for being a test anyway. 

Or am I missing something?
